Fix #4244: Patch source map links in the output of {fast,full}OptJS

gzm0 opened the pull request:
As a bonus, this allows us to quite easily test more cases. It is left as an exercise to the reader to spot the bug that got fixed on the way.

sjrd commented on linker-interface/shared/src/main/scala/org/scalajs/linker/interface/LegacyV1_2Adapter.scala:
```diff
@@ -0,0 +1,213 @@
+
+      res.append(reMatch.before)
+
+      /* If there is no source map link, keep an empty line to not break a
+       * potential (unlinked) source map
+       */
+      newLine.foreach(res.append(_))
```
It seems that the code does not in fact keep an empty line when `newLine.isEmpty`.

sjrd commented on linker-interface/shared/src/test/scala/org/scalajs/linker/interface/LegacyV1_2AdapterTest.scala:
```diff
@@ -0,0 +1,189 @@
+      .withJSFileURI(new URI("http://example.org/my-js-file-uri"))
+    
+    val readOut = new ReadOnlyOutputDirectory(
+        "main.js" -> raw"""
+        |console.log("hello");
+        |""".stripMargin,
+        "main.js.map" -> raw"""{
+        |  "other key": 1
+        |}""".stripMargin)
```
The indentation in these examples is confusing. It's not clear where elements start and where they end. Perhaps indent it as follows:
```suggestion
        "main.js" -> raw"""
          |console.log("hello");
          |""".stripMargin,
        "main.js.map" -> raw"""{
          |  "other key": 1
          |}""".stripMargin)
```
?

sjrd commented on linker-interface/shared/src/test/scala/org/scalajs/linker/interface/LegacyV1_2AdapterTest.scala:
```diff
@@ -0,0 +1,189 @@
+          |console.log("hello");
+          |//# sourceMappingURL=http://example.org/my-source-map-uri
+          |// some other comment
+          |""".stripMargin, writeOut.content("js"))
```
```suggestion
          |""".stripMargin,
          writeOut.content("js"))
```
(the first param spans several lines, so it should be alone on its lines)

The same applies in all the `assertEquals`.

sjrd commented on sbt-plugin/src/main/scala/org/scalajs/sbtplugin/ScalaJSPluginInternal.scala:
```diff
@@ -249,57 +249,51 @@ private[sbtplugin] object ScalaJSPluginInternal {
+        val outputJSFile = (artifactPath in legacyKey).value
+        val outputSourceMapFile = new File(outputJSFile.getPath + ".map")
+
+        IO.createDirectory(outputJSFile.getParentFile)
```
Is this line the hidden bug fix?

gzm0 replied:
Nope. That's also a bug fix indeed, but its untested :-/

gzm0 commented:
Updated. I believe I have addressed all the comments.

sjrd commented on linker-interface/shared/src/main/scala/org/scalajs/linker/interface/ReportToLinkerOutputAdapter.scala:
```diff
@@ -0,0 +1,211 @@
+import org.scalajs.logging.Logger
+
+import org.scalajs.linker.interface.unstable.{OutputDirectoryImpl, OutputFileImpl}
+
+
```
There's a double blank line here. :(

sjrd reviewed (approved):
LGTM, other than the double blank line.

gzm0 commented:
Well, I mean you also haven't found the bug yet, so we cannot possibly merge this yet anyways ;)

gzm0 commented:
Updated. In addition to removing the double blank line I have:
- Added a test case for no links before and after (completing the matrix of all cases).
- Removed trailing white space in `ReportToLinkerOutputAdapterTest`.

sjrd commented on sbt-plugin/src/main/scala/org/scalajs/sbtplugin/ScalaJSPluginInternal.scala:
```diff
@@ -249,57 +249,51 @@ private[sbtplugin] object ScalaJSPluginInternal {
+                .withSourceMapURI(relURI(outputSourceMapFile.getName()))
+                .withJSFileURI(relURI(outputJSFile.getName()))
+            }
+    
```
The CI says there is whitespace at EOL here.

sjrd commented on linker-interface/shared/src/main/scala/org/scalajs/linker/interface/ReportToLinkerOutputAdapter.scala:
```diff
@@ -0,0 +1,210 @@
+      implicit ec: ExecutionContext): Future[Unit] = {
+    retrieveOutputFiles(report, outputDirectory).flatMap {
+      case (jsFileContent, optSourceMapContent) =>
+
```
And it's not a big deal, but since there is something else to change anyway: this blank line is not supposed to be there, ideally.

gzm0 commented:
updated

sjrd commented:
Hum at https://scala-webapps.epfl.ch/sjsjenkins/blue/organizations/jenkins/scalajs-main/detail/PR-4247/3/pipeline/62 the 3 bootstrap test jobs seem to be stuck at the `testSuite2_1x/bootstrap:test`. If it were only 1 of them it could be spurious, but all 3 probably means that there is an actual issue.

gzm0 commented:
Yep, I can reproduce this locally. Seems like `convert` hangs when trying to read the output of `testSuiteLinker` (there is still a remaining `scalaJSLinkedFile` in the build). I'm investigating.

gzm0 commented:
There was a problem in the implementation of `readFull` in `PathOutputDirectory`. It is fixed + test added. Further, the failure behavior of `ReportToLinkerOutputAdapter` in case the source map does not contain a JSON object has been improved.

sjrd commented on linker/jvm/src/main/scala/org/scalajs/linker/PathOutputDirectory.scala:
```diff
@@ -100,6 +103,35 @@ object PathOutputDirectory {
+    object Handler extends CompletionHandler[Integer, Unit] {
+      def completed(read: Integer, unit: Unit): Unit = {
+        if (read == -1 || !buf.hasRemaining())
+          promise.success(buf.flip())
```
This does not compile on JDK 8:
```
[error] /localhome/jenkins/sjsjenkins/workspace/scalajs-main_PR-4247@2/linker/jvm/src/main/scala/org/scalajs/linker/PathOutputDirectory.scala:122:35: type mismatch;
[error]  found   : java.nio.Buffer
[error]  required: java.nio.ByteBuffer
[error]           promise.success(buf.flip())
[error]                                   ^
[error] one error found
```
`flip()` did not yet have an override in `ByteBuffer` to return a `ByteBuffer`. Instead we receive the one from `Buffer` which returns a `Buffer`. Either cast it down:
```suggestion
          promise.success(buf.flip().asInstanceOf[ByteBuffer])
```
or use a block:
```suggestion
          promise.success({ buf.flip(); buf })
```

gzm0 commented:
Wow... This one really demands everything :P
